File: lab03b.py
#!/usr/bin/env python3
"""
Programming Assignment 3: A simple interactive game
"""


def them(stones, start):
    """
    Human takes a turn.
    """
    print('There are', stones, 'stones')
    while True:
        take = int(input('How many do you take? '))
        if take == 0 and stones == start:
            print('In that case I will go first')
            break
        elif take < 1:
            print('You must take a stone')
        elif take > 3:
            print('You cannot take more than 3 stones')
        elif take > stones:
            print('There are only', stones, 'stones')
        else:
            break
    return take


def us(stones):
    """
    Computer takes a turn.
    """
    take = max(1, (stones - 1) % 4)  # must take at least one
    print('I take', take)
    return take


def play():
    """
    Play the game.
    """
    start = 21
    stones = start

    while stones > 0:
        stones -= them(stones, start)
        if stones == 0:
            print('You took the last stone. You lose.')
        else:
            stones -= us(stones)
            if stones == 0:
                print('I took the last stone. I lose.')


play()

File: lab03b.cpp
// Programming Assignment 3: A simple interactive game

/*
Pick-up-stones is a simple game to play and implement. Here's a simple
explanation of the game: Twenty-one stones are put in a pile and players take
turns removing 1, 2 or 3 stones from the pile. The last player to pick up a
stone from the pile loses. Even though it sounds easy to play, it might not be
easy to win every time you play. Your task is to implement this game so that you
can play against the computer (your program). Your program should employ some
clever algorithm so that it could win against a naive player. It does not have
to win all the time in order to get most of the credit. However, you will get
full credit if you write a program that is impossible to beat.

Hint: In order to make your program win all the time, the other player must go
first.
*/

#include <iostream>
#include <cstdlib>
#include <ctime>

using namespace std;

int main() {
    int stones = 21; // Initial number of stones
    int playerMove, computerMove;

    cout << "Welcome to the Pick-up-stones game!" << endl;
    cout << "There are 21 stones in the pile." << endl;
    cout << "You can pick 1, 2, or 3 stones on your turn." << endl;
    cout << "The player who picks the last stone loses." << endl;
    cout << "You will go first." << endl;

    while (stones > 0) {
        // Player's turn
        do {
            cout << "There are " << stones << " stones left." << endl;
            cout << "How many stones do you want to pick (1, 2, or 3)? ";
            cin >> playerMove;
        } while (playerMove < 1 || playerMove > 3 || playerMove > stones);

        stones -= playerMove;
        if (stones == 0) {
            cout << "You picked the last stone. You lose!" << endl;
            break;
        }

        // Computer's turn
        computerMove = (4 - playerMove) % 4; // Optimal move
        if (computerMove == 0) computerMove = 1; // If player picks 3, pick 1

        stones -= computerMove;
        cout << "Computer picks " << computerMove << " stone(s)." << endl;

        if (stones == 0) {
            cout << "Computer picked the last stone. Computer loses!" << endl;
            break;
        }
    }

    return 0;
}
